### DISCUSSION TITLE
UITableView의 높이가 스스로의 content만큼 되게 하는 방법

### DISCUSSION THREAD

**Discussion by JuTaK97** [wafflestudio/seminar-2023#173] [Q&A]
`private lazy var tableView: UITableView` 를 만들고,  `view.addSubview(tableView)` 한다음에 
```
       tableView.snp.makeConstraints { make in
            make.horizontalEdges.equalToSuperview()
            make.verticalEdges.equalToSuperview()
        }
```
이렇게 하면 전체 화면이 TableView로 꽉 차게 됩니다.

그런데 과제를 위해 TableView의 높이는 하단까지 꽉 차는게 아니라 높이가 TableView의 높이만큼으로 딱 맞춰졌으면 좋겠는데 이게 잘 안되네요

```
        tableView.rowHeight = UITableView.automaticDimension
        tableView.estimatedRowHeight = 50
```
대부분 이걸 하라고 알려주는데, 이건 하나마나 달라지는 게 없습니다. (똑같이 화면 꽉 참)
이 옵션은 셀 별로 높이가 달라질 때 그걸 동적으로 조절해 주는 거지, TableView 전체 높이와 관련된 게 아닌 것 같기도 하네요

그렇다고 snp.makeConstraints 쪽을 건드리면, 대부분 그냥 검정 화면만 꽉 차게 됩니다. (아무것도 안 뜨게 되는 거죠?)

안드로이드의 wrap_content 같은 걸 하려면 snp.makeConstraints를 어떻게 줘야되나요?

tableView.backgroundColor = .red 를 줬을 때

<img width="435" alt="image" src="https://github.com/wafflestudio/seminar-2023/assets/88367636/cdea6c72-af1c-4795-bb28-76577a5f736e">

빨간 부분이 아예 안 보이게 하고 싶습니다.(딱 content만큼의 size를 갖도록)

**Comment by shp7724** [comment]
[이런](https://stackoverflow.com/a/33268512/9046780) 식으로 뷰 라이프사이클 메서드에서 tableView height constraint를 업데이트를 할 수는 있을 것 같습니다. 

그런데 구현하고 싶은 UI가 따로 있는건가요? 과제 1은 위와 같은 workaround를 사용하지 않아도 완료할 수 있을 것 같아서요.

**Reply by JuTaK97** [reply]
아이템이 없을 때 UITableView 는 아예 크기가 없어져야 "미리 알림 없음"이 그 자리에 들어가는데, 지금 TableView를 equalToSuperview로 해놔서 그게 안됐어요

**Reply by JuTaK97** [reply]
snp.makeConstraints 에서 생각한 대로 동작하는게 지금 make.horizontalEdges.equalToSuperview() 랑 equalTo(숫자) 밖에 없는데, 보통 뭐를 주로 쓰나요?

**Reply by JuTaK97** [reply]
안드로이드의 wrap_content가 그리워요  😠😠😠😠😠😠

**Reply by shp7724** [reply]
아하.. 미리 알림 없음 화면을 tableView 안에 넣지 말고, 그 위에 띄우는 방법은 어떨까요? tableView가 비어 있는 상황에서만 해당 뷰를 보여주면 될 것 같아요.

**Reply by shp7724** [reply]
(아래는 제 개인적인 생각..)
만약 tableView의 높이와 content의 높이를 일치시켜야 하는 경우가 있다면 차라리 UIStackView를 사용하는 게 좋을 것 같아요.
UITableView는 UIScrollView 이기도 하기 때문에, content와 높이를 일치시킨다면 이 테이블뷰를 또 스크롤 뷰 안에 넣어야 할 것 같은데, 이렇게 하는 건 거의 못 본 것 같네요..

**Reply by shp7724** [reply]
`모든 미리 알림이 완료됨`의 위치가 스크롤 위치에 따라 바뀌지 않는 걸 보니 미리알림 앱도 비슷한 방식으로 구현된 것 같아요!

https://github.com/wafflestudio/seminar-2023/assets/33917774/ca9a497e-68af-4864-815c-a496da7a213f

**Reply by JuTaK97** [reply]
그러면 일반적으로 모든 UIView가 내부 content에 따라 크기가 바뀌어야 할 때는 UIStackView를 쓰는건가요?
snp.makeConstraints 수준에서 설정할 수 있으면 참 편할 것 같은데말이죠...

**Reply by shp7724** [reply]
음.. 상황마다 접근 방법이 너무 달라서 일반화해서 설명하기가 어렵네요 😅
일반적으로, 모든 UIView에는 `intrinsicContentSize`라는 프로퍼티가 존재하는데, 이것이 말씀하신 `내부 content`의 크기에 대응하는 개념입니다.
이 `intrinsicContentSize` 프로퍼티를 오버라이드하면 원하는 값을 유동적으로 return 하도록 할 수 있어요.
참고로 Auto Layout은 레이아웃을 계산할 때 이 `intrinsicContentSize`를 고려하는데요. 예를 들어, 어떤 뷰에 `intrinsicContentSize`가 정의되어 있다면, `leading`, `top` anchor만 잡아주어도 문제 없이 제약 조건을 계산할 수 있습니다.

관련 블로그 글 첨부합니다!
- https://babbab2.tistory.com/135

`intrinsicContentSize`를 커스터마이즈하는건 최후의 선택이고, 많은 경우 더 깔끔하게 처리할 수 있는 방법이 있는 것 같아요. 이건 나중에 구체적인 케이스와 함께 질문주시면 답변드릴게용
